### ISSUE TITLE
Generate Package-lock.json for every package

### ISSUE THREAD

**Issue by Vishesh30** [lerna/lerna#2417]
## Expected Behavior
Expected behaviour is to have a package-lock.json file generated for every package in packages folder.

## Current Behavior
My current project structure look like:-

packages/internal-package-1/package.json
packages/internal-package-2/package.json
packages/internal-package-3/package.json
lerna.json
package.json
package-lock.json

I am using lerna version 3.17.0.

Right now as shown above there is only one package-lock.json file which is generated for . the entire project and it only contains the dependency which in top package.json file.

My expectation was that for every package.json file corresponding package-lock.json should be generated but that is not the case.  Furthermore, the top package-lock.json file only contains the dependencies in the top package.json and not the all the dependencies which are declared in evey package.json file.

Now, if we try to consume for example internal-package-1 in a different project that as there is no lock file for this package , latest version of the dependencies gets downloaded which is not the expected behaviour.

## Possible Solution
Possible solution or expectation is to have a lock file generated for every package.

<details><summary>lerna.json</summary><p>

```json
{
  "packages": [
    "packages/*",
    "packages/Foundation/src/SampleNestedModule"    
  ],
  "version": "0.0.0"
}
```
</p></details>

## Context
This issue is affecting us because as the lock file is not generated for every package and if i try to consume the internal-package-1 in a different project then locked dependency are not getting downloaded but the latest version of them gets downloaded.

## Your Environment
We are hoisting the dependency hence we have modified out npm install script as below:-
"install": "lerna bootstrap --hoist"  , this correctly hoists the dependency but does not generate the lock file for individual package. 

| Executable | Version |
| ---: | :--- |
| `lerna --version` |3.17.0|
| `npm --version`  | 6.10.1 |
| `yarn --version` | Not using yarn |
| `node --version` | 10.16.0 |

| OS | Version |
MACOS
| NAME | VERSION |
| macOS Catalina | 10.15.2 |
-->

**Comment by Vishesh30** [comment]
I have tried looking into various questions around the same issue but got find a working solution ex:- https://github.com/lerna/lerna/issues/1462

**Comment by revelt** [comment]
@Vishesh30 please clarify, did #1462 posts help to get `package-lock` files? I personally [don't use](https://github.com/sindresorhus/ama/issues/479) lockfiles but I see unit tests in Lerna mentioning them: https://github.com/lerna/lerna/blob/2612f51e7eecec58eacf0571724e6989e4b8e42d/commands/version/__tests__/version-command.test.js#L680 

Let us know if the ticket can be closed.

**Comment by jannikbuschke** [comment]
Im seeing this issue too. I want to have package-lock.json files per package, as if I had run `npm install` in each. Using `bootstrap` without any arguments seems to yield this result. Using `bootstrap --hoist` this is not the case.

**Comment by Vishesh30** [comment]
> @Vishesh30 please clarify, did #1462 posts help to get `package-lock` files? I personally [don't use](https://github.com/sindresorhus/ama/issues/479) lockfiles but I see unit tests in Lerna mentioning them:
> 
> https://github.com/lerna/lerna/blob/2612f51e7eecec58eacf0571724e6989e4b8e42d/commands/version/__tests__/version-command.test.js#L680
> 
> Let us know if the ticket can be closed.

@revelt Nope, That does not work. I am still struggling to get a solution for this

**Comment by Vishesh30** [comment]
> Im seeing this issue too. I want to have package-lock.json files per package, as if I had run `npm install` in each. Using `bootstrap` without any arguments seems to yield this result. Using `bootstrap --hoist` this is not the case.
> 
> So it might just be an issue with hoisting?

@jannikbuschke  - But, If you don't use hoist then the install time would go up like anything as it would install all dependencies in all package which is not correct at all . I would not consider this as a solution.

**Comment by jannikbuschke** [comment]
@Vishesh30 yes agree, I just wanted to record that I *think* that bootstrap **without** `--hoist` works as expected (updates all packages/package-lock.json files) whereas bootstrap **with** `--hoist` does not.
This was not 100% clear from your description and could help the maintainers or whoever wants to tackle this help to understand/resolve the problem.

**Comment by Vishesh30** [comment]
Hi All,

I couldn't find a concrete solution to generate lock files for all packages. I mean there are ways but, everything is increasing the installing time to very high. Below are 2 ways to generate package-lock.json file for all packages:-

1. Directly use lerna bootstrap without --hoist flag ------- This does generate lock file but increases the install time way to high.
2. Use "lerna exec -- npm i" ------ This will generate the lock file but "install" times are way higher not a viable solution with 25 packages in my repository.


As above 2 solutions were taking way to much time hence i considered them as not a feasible solution for large repos hence, i came up with a third way or i would call it a workaround , this is also not the cleaneast  solution but does the job with very slight increase in installation time.

1) Create a npm script in all your packages which would generate only package-lock file without installation which would be something like below:-
           ` "genPackagelock": "npm i --package-lock-only"`

2) In you root package.json file as part of postinstall call the above defined script for all the packages as below:-
             `"postinstall": "lerna run --parallel createPackage-lock"`

The above "postinstall" basically generates package-lock.json file for all the packages along with the internal dependencies.

I am not closing the Bug because the above solutions are a workaround and not one of the best solution.

**Comment by Vishesh30** [comment]
I couldn't find a concrete solution to generate lock files for all packages. I mean there are ways but, everything is increasing the installing time to very high. Below are 2 ways to generate package-lock.json file for all packages:-

1. Directly use lerna bootstrap without --hoist flag ------- This does generate lock file but increases the install time way to high.
2. Use "lerna exec -- npm i" ------ This will generate the lock file but "install" times are way higher not a viable solution with 25 packages in my repository.


As above 2 solutions were taking way to much time hence i considered them as not a feasible solution for large repos hence, i came up with a third way or i would call it a workaround , this is also not the cleaneast  solution but does the job with very slight increase in installation time.

1) Create a npm script in all your packages which would generate only package-lock file without installation which would be something like below:-
           ` "genPackagelock": "npm i --package-lock-only"`

2) In you root package.json file as part of postinstall call the above defined script for all the packages as below:-
             `"postinstall": "lerna run --parallel genPackagelock"`

The above "postinstall" basically generates package-lock.json file for all the packages along with the internal dependencies.

I am not closing the Bug because the above solutions are a workaround and not one of the best solution.

@jannikbuschke @revelt  -- What do you think of above solution any thought or shortcomings you see.?

**Comment by sempasha** [comment]
> Now, if we try to consume for example internal-package-1 in a different project that as there is no lock file for this package , latest version of the dependencies gets downloaded which is not the expected behaviour.

According to package-lock.json description at [npmjs.com](https://docs.npmjs.com/files/package-lock.json)

> One key detail about package-lock.json is that it cannot be published, and it will be ignored if found in any place other than the toplevel package

package-lock.json affects only top level package (i.e. end application, but not npm installed modules). That means package-lock.json required only for application packages (which usually won't published to registry at all, but installed directly from git instead), so there is no need to commit package-lock.json for packages which are libraries. So there may be a mistake to generate package-lock.json for every package controled by lerna.

**Comment by n370** [comment]
Seems to happen if you pass the "--scope" flag too. lerna@3.20.2

**Comment by AndrewSouthpaw** [comment]
Can confirm we have this problem too. One of the workarounds appears to be passing the `--force-local` flag, which gets around this execution branch that disables the lockfile generation:

https://github.com/lerna/lerna/blob/a47fc294393a3e9507a8207a5a2f07648a524722/commands/bootstrap/index.js#L149

(Hat tip to @liamuk for discovering.)

**Comment by JamesHenry** [comment]
Hi Folks,

This issue related to the legacy package management capabilities of Lerna. The world in which Lerna had to manage package installation and linking concerns because of gaps in the ecosystem no longer exists, we are blessed to have many great options now, not just `npm` (which has also massively improved since the early days).

In June 2023, in Lerna `7.0.0`, we removed the `lerna bootstrap`, `lerna add` and `lerna link` commands in Lerna by default. Over two years later, in September 2025 (Lerna `9.0.0`), the commands were finally fully removed.

We provided some guidance at the time on how to migrate existing codebases to use native package manager features here: https://lerna.js.org/docs/legacy-package-management

I'll close this one as no action is applicable, but please always feel free to open a fresh issue giving as much context as possible when using the latest and greatest and we'll take a look. I'm making sure we close older issues like this one to free up bandwidth to make sure we tackle what matters most to users in 2026.

Many thanks for your contributions to Lerna!
